state/storage.go

```diff
@@ -37,6 +37,22 @@ type LatestEvents struct {
 	LatestNID int64
 }
 
+// DiscardIgnoredMessages modifies the struct in-place, replacing the Timeline with
+// a copy that has all ignored events omitted. The order of timelines is preserved.
+func (e *LatestEvents) DiscardIgnoredMessages(shouldIgnore func(sender string) bool) {
+	// A little bit sad to be effectively doing a copy here---most of the time there
+	// won't be any messages to ignore (and the timeline is likely short). But that copy
+	// is unlikely to be a bottleneck.
+	newTimeline := make([]json.RawMessage, 0, len(e.Timeline))
+	for _, ev := range e.Timeline {
+		parsed := gjson.ParseBytes(ev)
+		if parsed.Get("state_key").Exists() || !shouldIgnore(parsed.Get("sender").Str) {
```

Comment:
It would be slightly faster to check shouldIgnore first and then if true, pull out the state key, as `ParseBytes` isn't cheap.

Reply:
Unfortunately, we need to parse the event to get the sender.